### PR TITLE
bulk-cdk: add ExceptionClassifier, revamp exception handling

### PR DESCRIPTION
**Pull request by postamar**
## What
This change is required for CAT tests to pass for connectors built using the Bulk CDK.
This change also paves the way to declarative rules-based exception handling.

Informs https://github.com/airbytehq/airbyte-internal-issues/issues/9373

## How
The change centers around the introduction of `ExceptionClassifier`

## Review guide
Reading through the diff in the order that github presents it is actually pretty good.

## User Impact
None

## Can this PR be safely reverted and rolled back?

- [x] YES 💚
- [ ] NO ❌

### CODE REVIEW

File: airbyte-cdk/bulk/core/base/src/main/kotlin/io/airbyte/cdk/AirbyteConnectorRunnable.kt [hunk 1]
```diff
@@ -15,18 +16,31 @@ private val log = KotlinLogging.logger {}
...
+
     override fun run() {
-        log.info { "Executing ${operation::class} operation." }
+        var operation: Operation? = null
         try {
+            try {
+                operation = operationProvider.get()!!
+            } catch (e: Throwable) {
+                throw ConfigErrorException("Failed to initialize connector operation", e)
+            }
```
Comment (postamar) [inline airbyte-cdk/bulk/core/base/src/main/kotlin/io/airbyte/cdk/AirbyteConnectorRunnable.kt:32]: This change now causes Micronaut's instantiation of `Operation` to deferred to this try-catch block which handles any exceptions as ConfigErrorExceptions. This is a requirement to pass the CAT tests, there's a case in there that deliberately attempts a read with an invalid configured catalog and expects a TRACE ERROR message.

File: airbyte-cdk/bulk/core/extract/src/main/kotlin/io/airbyte/cdk/check/CheckOperation.kt [hunk 2]
```diff
@@ -34,37 +37,15 @@ class CheckOperation<T : ConfigurationJsonObjectBase>(
...
-                    .withInternalMessage(e.toString())
-                    .withStackTrace(ExceptionUtils.getStackTrace(e)),
-            )
-            outputConsumer.accept(
-                AirbyteConnectionStatus()
-                    .withMessage(message)
-                    .withStatus(AirbyteConnectionStatus.Status.FAILED),
-            )
-            log.info { "Config check failed." }
-            return
```
Comment (postamar) [inline airbyte-cdk/bulk/core/extract/src/main/kotlin/io/airbyte/cdk/check/CheckOperation.kt:60]: All this `SQLException` logic had no business being in the general-purpose `CheckOperation` anyway. It's been moved to `JdbcExceptionClassifier` in the JDBC toolkit, where it belongs.

File: airbyte-cdk/bulk/core/base/src/main/kotlin/io/airbyte/cdk/AirbyteConnectorRunnable.kt (hunk 1 above)
Comment (akashkulk) [reply airbyte-cdk/bulk/core/base/src/main/kotlin/io/airbyte/cdk/AirbyteConnectorRunnable.kt:32]: Will this also mark system errors/other crashes as Config errors?

Comment (postamar) [reply airbyte-cdk/bulk/core/base/src/main/kotlin/io/airbyte/cdk/AirbyteConnectorRunnable.kt:32]: Yes in theory. In practice, exceptions thrown at this point will be triggered by the CLI inputs exclusively.

File: airbyte-cdk/bulk/toolkits/extract-jdbc/src/main/kotlin/io/airbyte/cdk/output/JdbcExceptionClassifier.kt [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+package io.airbyte.cdk.output
+
+import io.airbyte.cdk.Operation
+import io.micronaut.context.annotation.Replaces
+import io.micronaut.context.annotation.Value
+import java.sql.SQLException
+
+@Replaces(ExceptionClassifier::class)
+class JdbcExceptionClassifier(@Value("\${${Operation.PROPERTY}}") operationName: String) :
```
Comment (akashkulk) [inline airbyte-cdk/bulk/toolkits/extract-jdbc/src/main/kotlin/io/airbyte/cdk/output/JdbcExceptionClassifier.kt:13]: Would extending this class in the connectors be the way to implement the error framework that @theyueli added  e.g.  `PostgresSourceExceptionHandler`?

**Review by akashkulk — CHANGES_REQUESTED** [review]
Some quick questions

File: airbyte-cdk/bulk/toolkits/extract-jdbc/src/main/kotlin/io/airbyte/cdk/output/JdbcExceptionClassifier.kt (hunk 3 above)
Comment (postamar) [reply airbyte-cdk/bulk/toolkits/extract-jdbc/src/main/kotlin/io/airbyte/cdk/output/JdbcExceptionClassifier.kt:13]: Yes and no. This class such as it is exists only to maintain existing functionality, which is pretty bare bones. I didn't want to add too much stuff to this PR. However, it's not really what we want; what we want is a bunch of classifiers among which @theyueli 's (the one with `ConnectorErrorProfile`s) would be one of them.

I see adding functionality as the logical next step. Micronaut really unlocks a lot of possibilities here, it should be trivial to add some kind of rule definitions into a yaml file such as https://github.com/airbytehq/airbyte-enterprise/blob/master/airbyte-integrations/connectors/source-firetruck/src/main/resources/application.yml

The obvious first choice is to have some kind of rules engine that works on SQLException vendor codes.

**Comment by postamar** [comment]
Thanks for taking a look!

**Review by akashkulk — APPROVED** [review]

**Pull request merged**